harvdev_utils/chado_functions/feature.py

```diff
@@ -0,0 +1,313 @@
+
+    Lookup the feature by uniquename and by symbol and make sure this is the same.
+
+    features are unique wrt:- uniquename, organism_id and type_id
+    So we need to make sure we have all 3 to be safe. (though normally just the uniquename may do)
+
+    uniquename : FBxx0000001 type. Also be aware of things like FBgn0000014:11 which is an exon.
+
+    .. note::
+        If user supplies no type_name or organism, we can get the probable organism form the synonym
+        then lookup just using unique name and report problem if more than one found.
+
+    Args:
+        session (sqlalchemy.orm.session.Session object): db connection  to use.
+
+        uniquename (str): feature uniquename.
+
+        synonym (str): symbol to look up.
+
+        type_name (str) : <optional> cvterm name for the type of feature.
+
+        organism_id (int): <optional> chado organism_id.
+    Returns:
+        Feature object.
+
+    Raises:
+        DataError: if feature cannot be found uniquely.
+    """
+    try:
+        feature = get_feature_by_uniquename(session, uniquename, type_name=type_name, organism_id=organism_id)
+    except NoResultFound:
+        message = "Unable to find Feature with uniquename {}.".format(uniquename)
+        raise DataError(message)
+    except MultipleResultsFound:
+        message = "Found more than feature with this 'uniquename' {}.".format(uniquename)
+        raise DataError(message)
+
+    try:
+        feat_check = feature_symbol_lookup(session, type_name, synonym)
+    except NoResultFound:
+        message = "Unable to find Feature with symbol {}.".format(synonym)
+        raise DataError(message)
+    except MultipleResultsFound:
+        message = "Found more than feature with this symbol {}.".format(synonym)
+        raise DataError(message)
+
+    if feat_check.feature_id != feature.feature_id:
+        message = "Symbol {} does not match that for {}.".format(synonym, uniquename)
+        raise DataError(message)
+    return feature
+
+
+def feature_name_lookup(session, name, organism_id=None, type_name=None, type_id=None):
+    """Get feature by its name.
+
+    Lookup feature using the feature name.
+
+    If no organism_id is given then it will default to Dmel
+
+    Args:
+        session (sqlalchemy.orm.session.Session object): db connection  to use.
+
+        name (str): feature name.
+
+        type_name (str) : <optional> cvterm name for the type of feature.
+
+        organism_id (int): <optional> chado organism_id.
+
+    Returns:
+        Feature object.
+
+    Raises:
+        DataError: if feature not found uniquely.
+    """
+    if type_name and type_id:
+        raise CodingError("Cannot specify type_name and type_id")
+
+    # Default to Dros if no organism specified.
+    if not organism_id:
+        organism_id = get_default_organism_id(session)
+
+    if type_name:
+        if type_name in ['bogus symbol', 'single balancer', 'chemical entity', 'disease implicated variant']:
```

Comment:
Is this exhaustive? Are there others?

Reply:
Maybe it's better to do a cv term lookup here and find the appropriate cv? Or perhaps that might be too slow? Just worried about forgetting to add to this list at some point in the future.

Reply:
Wondering if we can just lookup the cv term for both the FB misc CV and the SO and then pick the appropiate cv based on where the cv term matches up successfully. If that makes sense.

Reply:
Then it's not hardcoded.

Reply:
cvterm is not unique enough. select * from cvterm where name = 'gene';
 cvterm_id | cv_id |                                    definition                                    | dbxref_id | is_obsolete | is_relationshiptype | name 
-----------+-------+----------------------------------------------------------------------------------+-----------+-------------+---------------------+------
       219 |     6 |                                                                                  |       191 |           0 |                   0 | gene
     57137 |    13 | Symbol of the gene corresponding to a sequence region (usable with all features) |   2413339 |           0 |                   0 | gene
    103415 |    52 | interactor for genetic interaction.                                              |   6381219 |           0 |                   0 | gene
(3 rows)

Ah okay lookup for both, slower but safer I guess.

Reply:
But could you just do the lookup with CV as SO and the lookup with CV as FB misc CV? To narrow down the search? And then pick whichever CV it hit successfully?

Reply:
Ah, missed your last line, yes, that's what I was thinking. Slower but safer. If the performance hit is too much, we can scrap it. But I think it might be nicer than hardcoding the data.